### PR TITLE
[SPARK-46145][SQL] spark.catalog.listTables does not throw exception when the table or view is not found

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

There is a race condition case when using `spark.catalog.listTables`. It happens when one client is calling `listTables` while another client is delete some of the tables. In this case `listTables` might throw `TABLE_OR_VIEW_NOT_FOUND` when it tries to access a table or view that is deleted concurrently. 

This PR proposes to improve `spark.catalog.listTables` not throw the exception but return normally.

### Why are the changes needed?

This is to improve user experience when using `spark.catalog.listTables` under the concurrent case. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### Was this patch authored or co-authored using generative AI tooling?

NO

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 1]
```diff
@@ -146,38 +147,44 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
...
-    val catalog = qe.analyzed.collectFirst {
-      case ShowTables(r: ResolvedNamespace, _, _) => r.catalog
-      case _: ShowTablesCommand =>
-        sparkSession.sessionState.catalogManager.v2SessionCatalog
-    }.get
-    val tables = qe.toRdd.collect().map { row =>
-      val tableName = row.getString(1)
-      val namespaceName = row.getString(0)
-      val isTemp = row.getBoolean(2)
+  private[sql] def resolveTable(row: InternalRow, catalog: CatalogPlugin): Option[Table] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:150]: it seems we only need to pass in the catalog name

File: sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala [hunk 2]
```diff
@@ -1050,6 +1052,16 @@ class CatalogSuite extends SharedSparkSession with AnalysisTest with BeforeAndAf
       func2.className.startsWith("org.apache.spark.sql.internal.CatalogSuite"))
   }
 
+  test("SPARK-46145: listTables does not throw exception when the table or view is not found") {
+    val impl = spark.catalog.asInstanceOf[CatalogImpl]
+    for (temp <- Seq(true, false)) {
+      val row =
+        new GenericInternalRow(Array(UTF8String.fromString("db"), UTF8String.fromString("t"), temp))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala:1059]: shall we make the name more explicit? `non_existing`

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala:1059]: and let's set schema to empty string for temp views.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 1 above)
Comment (amaliujia) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: right

File: sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala (hunk 2 above)
Comment (amaliujia) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala [hunk 3]
```diff
@@ -1050,6 +1052,15 @@ class CatalogSuite extends SharedSparkSession with AnalysisTest with BeforeAndAf
       func2.className.startsWith("org.apache.spark.sql.internal.CatalogSuite"))
   }
 
+  test("SPARK-46145: listTables does not throw exception when the table or view is not found") {
+    val impl = spark.catalog.asInstanceOf[CatalogImpl]
+    for (temp <- Seq((true, ""), (false, "non_existing_db"))) {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala:1057]:
```suggestion
    for ((isTemp, dbName) <- Seq((true, ""), (false, "non_existing_db"))) {
```

**Review by cloud-fan — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 4]
```diff
@@ -146,38 +147,44 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
...
+  private def makeTablesDataset(plan: ShowTables): Dataset[Table] = {
+    val qe = sparkSession.sessionState.executePlan(plan)
+    val catalog = qe.analyzed.collectFirst {
+      case ShowTables(r: ResolvedNamespace, _, _) => r.catalog
+      case _: ShowTablesCommand =>
+        sparkSession.sessionState.catalogManager.v2SessionCatalog
+    }.get
+    val tables = qe.toRdd.collect().flatMap { row => resolveTable(row, catalog.name()) }
     CatalogImpl.makeDataset(tables.toImmutableArraySeq, sparkSession)
   }
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:189]: Shall you move makeTablesDataset above resolveTable ?
So the change is less and easier to review.

Comment (amaliujia) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: done

**Review by beliefer — APPROVED** [review]
LGTM.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**